### PR TITLE
Support different agent modes in the native-image Gradle plugin

### PR DESCRIPTION
**Pull request by gradinac**
This PR reworks agent support by decoupling the agent task instrumentation from image building and introduces the concept of agent modes.

### CODE REVIEW

File: native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java [hunk 1]
```diff
@@ -64,6 +67,14 @@ public interface GraalVMExtension {
      */
     Property<Boolean> getTestSupport();
 
+    @Nested
+    AgentOptions getAgent();
+
+    @Nested
+    CopyMetadataOptions getCopyMetadata();
```
Comment (melix) [inline native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java:74]: this should probably live under `AgentOptions`

File: samples/java-application-with-reflection/settings.gradle [hunk 2]
```diff
@@ -45,6 +45,7 @@ pluginManagement {
     }
     repositories {
         mavenCentral()
+        mavenLocal()
```
Comment (melix) [inline samples/java-application-with-reflection/settings.gradle:48]: Will need to remove this from draft PR

File: native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java (hunk 1 above)
Comment (gradinac) [reply native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java]: I agree, I've moved this to `AgentOptions` and renamed it to `metadataCopy` to remain consistent with the naming of other tasks

File: samples/java-application-with-reflection/settings.gradle (hunk 2 above)
Comment (gradinac) [reply samples/java-application-with-reflection/settings.gradle:48]: Ah yes, I've kept this to make it easier to test locally on multiple machines, I will remove all changes related to local testing

File: samples/java-application-with-reflection/build.gradle [hunk 3]
```diff
@@ -46,6 +46,7 @@ plugins {
 
 repositories {
     mavenCentral()
+    mavenLocal()
```
Comment (melix) [inline samples/java-application-with-reflection/build.gradle:49]: Should be removed

**Review by melix — COMMENTED** [review]
This is looking good, let's get this to the finish line!

File: native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/agent/AgentOptions.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Input
+    ListProperty<String> getFilterableEntries();
+
+    /**
+     * Configures the tasks which needs to be instrumented
+     *
+     * @return .
+     */
+    @Input
```
[the remark this answers is not in the document]
Comment (gradinac) [reply native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/agent/AgentOptions.java:91]: That's a good point, I'll change it to `@Internal`

File: native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/actions/CleanupAgentFilesAction.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.gradle.api.Action;
+import org.gradle.api.Task;
+import org.gradle.api.file.FileSystemOperations;
+import org.gradle.api.provider.Provider;
+
+import java.util.List;
+
+public class CleanupAgentFilesAction implements Action<Task> {
+
+    private final Provider<List<String>> directoriesToCleanup;
```
[the remark this answers is not in the document]
Comment (gradinac) [reply native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/actions/CleanupAgentFilesAction.java:52]: Hmm, we could introduce a `ListProperty`, but we would manually have to instantiate it and pass it around. We could also go with a `ConfigurableFileCollection`, though in this context these files are either passed in from a task input, or created internally by mapping another property

File: native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/agent/AgentOptions.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (melix) [reply native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/agent/AgentOptions.java:143]: This is the last remaining issue AFAICT.

**Review by melix — APPROVED** [review]
Given that the "conditional" mode is the recommended mode for libraries, I wonder if we should make it the default when the `java-library` plugin is applied?

File: native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/agent/AgentOptions.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gradinac) [reply native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/agent/AgentOptions.java:143]: Snap, sorry for that, I missed this when refactoring, will push an updated version

**Comment by gradinac** [comment]
> Given that the "conditional" mode is the recommended mode for libraries, I wonder if we should make it the default when the `java-library` plugin is applied?

That's a good point - we could do this if the agent mode is not explicitly set on the command line and the user has not specified a `defaultMode in the DSL.

I've noticed plugins behave very similar to tasks in terms of lazyness - there's a `withType` call that's called back only when a plugin actually executes. For the agent, we hook into project evaluation so I don't think we can use that mechanism.
The alternative is to simply check if the `JavaLibraryPlugin` is included in the project.

One downside of doing this would be non-apparent changes in the build configuration that depend on included plugins

File: native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/actions/CleanupAgentFilesAction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (melix) [reply native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/actions/CleanupAgentFilesAction.java:52]: Let's keep it as is for now, I'll take a look when I have time

**Comment by melix** [comment]
mmm, not sure I follow you. You can do:

```
project.getPlugins().withType(JavaLibraryPlugin.class) { configure agent convention }
```

**Comment by gradinac** [comment]
> mmm, not sure I follow you. You can do:
> 
> ```
> project.getPlugins().withType(JavaLibraryPlugin.class) { configure agent convention }
> ```

Ah yes, this would set the convention, but that's the bit i'm unsure about - would this execute before or after project evaluation?

**Comment by melix** [comment]
It would execute as soon as the `java-library` plugin is applied. That is to say immediately if it's already applied when this code executes, or later if it's not yet applied.

**Comment by gradinac** [comment]
> It would execute as soon as the `java-library` plugin is applied. That is to say immediately if it's already applied when this code executes, or later if it's not yet applied.

Makes sense, thank you! :) I'll push an update to the PR

**Review by melix — APPROVED** [review]

**Pull request merged**